apps/aecore/src/aec_sync.erl

```diff
@@ -278,59 +367,129 @@ do_forward_tx(Tx, Uri) ->
     do_get_missing_blocks(Uri),
     aec_events:publish(chain_sync, {server_done, Uri}).
 
-fetch_chain(Uri, GHash) ->
-    case aeu_requests:top(Uri) of
-        {ok, Hdr} ->
-            lager:debug("Top hdr (~p): ~p", [Uri, pp(Hdr)]),
-            {ok, Hash} = aec_headers:hash_header(Hdr),
-            fetch_chain(Hash, Uri, GHash, true, []);
-        {error, Reason} ->
-            lager:debug("fetching top block (~p) failed: ~p", [Uri, Reason])
-    end.
 
-%% Fetch the chain, keep track of if we can be connected to genesis.
-%% Initially we can be connected, later we can only be connected
-%% after we have actually fetched a block from another peer, and then
-%% found a block locally (in the extreme case that local block will
-%% be the genesis).
-fetch_chain(GHash, Uri, GHash, _MaybeConnected, Acc) ->
-    chain_fetched(Uri, Acc);
-fetch_chain(Hash, Uri, GHash, MaybeConnected, Acc) ->
-    case do_fetch_block(Hash, Uri) of
-        {ok, false, Block} -> %% We already have this block
-            case MaybeConnected andalso is_connected_to_genesis(Hash) of
-                true ->
-                    %% We are done
-                    chain_fetched(Uri, Acc);
-                false ->
-                    Hdr = aec_blocks:to_header(Block),
-                    fetch_chain(aec_headers:prev_hash(Hdr), Uri, GHash, false, Acc)
-            end;
-        {ok, true, Block} -> %% We fetched this block
-            Hdr = aec_blocks:to_header(Block),
-            fetch_chain(aec_headers:prev_hash(Hdr), Uri, GHash, true, [{Hash, Block}|Acc]);
-        {error, _} ->
-            %% Post the bit we have
-            chain_fetched(Uri, Acc)
+fetch_chain(Uri, FromHeight, ToHeight, _) when FromHeight == ToHeight ->
+    aec_events:publish(chain_sync, {client_done, Uri});
+fetch_chain(Uri, FromH, ToH, HashFromH) when FromH < ToH ->
+    case aeu_requests:get_header_by_height(Uri, FromH + 1) of
+        {ok, Header} ->
+          lager:debug("Header fetched (~p): ~p", [Uri, pp(Header)]),
+          case aec_headers:prev_hash(Header) == HashFromH of
+            true ->
+              {ok, HeaderHash} = aec_headers:hash_header(Header),
+              fetch_chain_from_header(Uri, FromH + 1, HeaderHash); 
+            false ->
+              lager:info("Abort sync due to non-fitting header ~p =/= ", [Header]),
+              delete_from_pool(Uri),
+              {error, sync_abort}
+          end;
+      {error, Reason} ->
+          delete_from_pool(Uri),
+          {error, Reason}
     end.
 
-is_connected_to_genesis(Hash) ->
-    lager:debug("Checking if ~p is connected to genesis", [pp(Hash)]),
-    aec_chain:hash_is_connected_to_genesis(Hash).
+fetch_chain_from_header(Uri, Height, HeaderHash) ->
+  case do_fetch_block(HeaderHash, Uri) of
+    {ok, true, Block} ->
+      lager:debug("Calling post_block(~p) --> hash of header ~p", [pp(Block), HeaderHash]),
+      case aec_conductor:add_synced_block(Block) of
```

Comment:
I see approach is similar to before in only posting blocks to conductor (I do not see calls to `aec_conductor:post_header/1`).